R/psis_approximate_posterior.R

```diff
@@ -0,0 +1,46 @@
+#' Diagnostics for Laplace and ADVI approximations and Laplace-loo and ADVI-loo
+#'
+#' @param log_p The log-posterior (target) evaluated at samples from the proposal distribution (q).
+#' @param log_q The log-density (proposal) evaluated at samples from the proposal distribution (q).
+#' @param log_liks A log-likelihood matrix. See \code{\link{loo.matrix}}. Default is \code{NULL}. Then only the posterior is evaluated using the k_hat diagnostic.
+#' @inheritParams loo
+#'
+#' @return
+#' If log likelihoods are supplied, the function returns a \code{loo} object,
+#' otherwise the function returns a \code{psis} object.
+#'
+#' @seealso \code{loo} and \code{psis}
+#'
+#' @template loo-and-psis-references
+#'
+#' @export
+psis_approximate_posterior <- function(log_p, log_q, log_liks = NULL, cores, save_psis, ...){
+  checkmate::assert_numeric(log_p, any.missing = FALSE, len = length(log_q))
+  checkmate::assert_numeric(log_q, any.missing = FALSE, len = length(log_p))
+  checkmate::assert_matrix(log_liks, null.ok = TRUE, nrows = length(log_p))
+  checkmate::assert_integerish(cores)
+  checkmate::assert_flag(save_psis)
+
+  approx_correction <- log_p - log_q
+
+  if(is.null(log_liks)){
+    approx_correction <- approx_correction - max(approx_correction) # Handle underflow/overflow
+    log_ratios <- matrix(approx_correction, ncol = 1)
+  } else {
+    log_ratios <- -log_liks
+    for(j in 1:ncol(log_ratios)){
+      log_ratios[,j] <- log_ratios[,j] + approx_correction
+      log_ratios[,j] <- log_ratios[,j] - max(log_ratios[,j]) # Handle underflow/overflow
+    }
+  }
+  psis_out <- psis.matrix(log_ratios = log_ratios, cores = cores, r_eff = rep(1, ncol(log_ratios)))
```

Comment:
I think you can replace `r_eff = rep(1, ncol(log_ratios))` with `r_eff = NA`. I think I enabled `NA` as a way to have it ignore the `r_eff` argument without throwing a warning. But no big deal either way here. 

```suggestion
  psis_out <- psis.matrix(log_ratios = log_ratios, cores = cores, r_eff = NA)
```

Reply:
Aki made the same comment and I tried it but it actually did not work. Maybe I'm in some old branch. I can do it, but then the current test suite will fail.

Just let me know how you want it.

Reply:
It's fine as is, don't worry about it.